### DISCUSSION TITLE
왜 값이 두번 들어가지? 왜 값이 두번 들어가지?

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#241] [질문있어요] [answered]
# 사전 설명

친애하는 백엔드 여러분 안녕하세요
저희가 설계한 스키마 구조를 먼저 설명하겠습니다.

저희는 히스토리의 카테고리를 분류하기 위해 HistoryCategory라는 엔티티를 생성하여 사용하고있습니다.
해당 값은 자바 객체 내부에서 Enum으로 관리되고 있습니다.

```java
@Getter
@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class HistoryCategory extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    @NotNull
    @Column(name = "history_type", nullable = false)
    private HistoryType historyType;

    @Builder
    private HistoryCategory(HistoryType historyType) {
        this.historyType = historyType;
    }
}
```

# 문제상황

저 같은 경우 PetPlantService를 테스트하는 과정에서 문제를 발견했습니다.

저희는 ReminderService를 구현하는 과정에서 다음과 같은 코드로 HistoryCategory를 찾아온 뒤 저장하게됩니다.

```java
HistoryCategory historyCategory = historyCategoryRepository.findByHistoryType(HistoryType.LAST_WATER_DATE)
                .orElseThrow(
                        () -> new NoSuchElementException("존재하지 않는 히스토리 타입입니다. type: " + HistoryType.LAST_WATER_DATE));
```

해당 코드는 HistoryType은 단 한개만 존재해야한다 (unique) 조건이 선행되어야 무결한 코드입니다.

![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/2f1e7852-3b7f-447e-b0e5-2004b9b3ad41)

때문에 HistoryType에 unique조건을 추가했습니다

![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/e8bd1de9-a3bd-4eb3-9e21-2c794a5b70a9)

그러자 테스트에서 다음과 같이 unique 제약조건으로 인한 문제가 터졌습니다.

확인 결과 `@BeforeEach`에서 값이 두번 씩 저장되고 있었습니다.

# 요약

`PetPlantServiceTest`의 `@BeforeEach`에서 값이 두번 씩 저장되고 있었습니다.

```java
// 문제의 코드 중 일부
@BeforeEach
void setUp() {
    dictionaryPlant = dictionaryPlantSupport.builder().build();
    member = memberSupport.builder().build();
    for (HistoryType type : HistoryType.values()) {
        historyCategorySupport.builder()
                .historyType(type)
                .build();
    }
}
```

따라서 궁금한 점은 왜 2번씩 저장되는가? 입니다

**Comment by Choi-JJunho** [comment]
해당 문제에 대한 원인을 파악해서 공유합니다.

# 원인

인수테스트에서 저장된 값이 통합테스트에 흘러들어옴.

인수테스트에서 컨텍스트 영역에 저장된 값이 통합테스트에서 초기화하지 못하여 발생하는 문제로 파악했습니다.

기존 통합테스트의 경우 `@Transactional`어노테이션을 이용해 테스트끼리 격리되도로 구성했습니다.
하지만 동일한 트랜잭션의 범위가 아닌 인수테스트의 경우 실제 DB에 값을 넣고 초기화 하는 방식을 사용하고 있으며 이 방식은 BeforeEachCallback으로 구현되어있어 `인수테스트가 실행되기 전`에만 DB를 초기화시킵니다.

결국 인수테스트가 다 끝난 뒤에는 데이터가 남아있게 되는 것이죠...

# 조치 방법

이에 다음과 같은 방법을 생각해볼 수 있습니다.

1. 인수테스트의 BeforeEach/AfterEach에 에 DB 초기화 구문을 작성한다. (이 경우 통합테스트의 `@Transactional`을 유지할 수 있다는 특징이 있습니다.)

2. 통합테스트의 DB 초기화 전략을 인수테스트와 동일하게 `DatabaseClearExtension`을 이용하는 방식을 적용한다.

**Reply by Kim0914** [reply]
`인수테스트에서 저장된 값이 통합테스트에 흘러들어옴` -> 이 상황은 히스토리 타입에만 적용되는 상황이 아니고, 전체 테스트에 적용되는 상황인거겠죠?!

**Reply by Choi-JJunho** [reply]
넵 히스토리 뿐만 아닌 `@Transactional`을 이용한 초기화 전략을 가져가고있는 다른 테스트에도 동일하게 적용되고 있습니다.

**Comment by Kim0914** [comment]
beforeEach에서 히스토리 카테고리를 생성하는데

반려_식물_삭제시_히스토리도_함께_삭제된다() 테스트에서 history를 생성하는 historySupport 내부에서 카테고리를 또 생성합니다.

이 부분도 고려해야겠네요 !

![image](https://github.com/woowacourse-teams/2023-pium/assets/68818952/685ca334-9ffd-456d-947a-b52d82b728bc)

### historySupport
![image](https://github.com/woowacourse-teams/2023-pium/assets/68818952/52a84d93-658b-41ef-bbb0-0bc58430595a)
historyCategorySupport가 생성하고 있어요 🧐

**Comment by yeonkkk** [comment]
> 1. 인수테스트의 BeforeEach/AfterEach에 에 DB 초기화 구문을 작성한다. (이 경우 통합테스트의 `@Transactional`을 유지할 수 있다는 특징이 있습니다.)

저는 첫번째 방법이 좋은 것 같아요.
2번의 경우 새로운 테스트가 생겨날 때마다 추가적으로 작업을 해줘야하기 때문에 그 만큼의 시간도 투자해야하고 놓치는 부분(예상하지 못한 부분)이 생길 수 있을 것 같아요.
1번으로 진행하면 상대적으로 적은 비용으로 전체 테스트 코드에 영향이 가는 것을 막을 수 있을 것 같습니다~

**Comment by Kim0914** [comment]
인수테스트에서 저장된 값이 통합테스트에 흘러들어오는 상황이 모든 테스트에 공통적으로 적용됩니다 !

그러므로 인수테스트의 BeforeEach/AfterEach에 에 DB 초기화 구문을 작성하는 방법을 선택하는 방안이 좀 더 나아보입니다.
여기서 DB 초기화 구문은 현재 테스트에 위치하고 있는 디비 클리너를 활용할 수 있을까요?
~~AfterEachCallback~~ `AfterAll`도 있는 것 같은데 활용할 수 있으면 활용해봐도 되겠네요 !

추가적으로, 인수 테스트의 이슈와는 별개로 위 코멘트에서 언급했던 부분도 해결해야할 것 같아요 😀
setUp에서 초기화 하지 말고, 사용하려고 하는 테스트 내부에서 생성하도록 하는 것은 어떤가요?!
그러면 굳이 히스토리 카테고리가 필요없는 테스트에서 생성되는 부분도 막을 수 있을 것 같아요 !

> beforeEach에서 히스토리 카테고리를 생성하는데
> 
> 반려_식물_삭제시_히스토리도_함께_삭제된다() 테스트에서 history를 생성하는 historySupport 내부에서 카테고리를 또 생성합니다.
> 
> 이 부분도 고려해야겠네요 !
> 
> ![image](https://user-images.githubusercontent.com/68818952/259624414-685ca334-9ffd-456d-947a-b52d82b728bc.png)
> 
> ### historySupport
> ![image](https://user-images.githubusercontent.com/68818952/259624635-52a84d93-658b-41ef-bbb0-0bc58430595a.png) historyCategorySupport가 생성하고 있어요 🧐

**Reply by Choi-JJunho** [reply]
> 여기서 DB 초기화 구문은 현재 테스트에 위치하고 있는 디비 클리너를 활용할 수 있을까요?

넵 활용 가능합니다!

> HistoryCategory에 대해 : setUp에서 초기화 하지 말고, 사용하려고 하는 테스트 내부에서 생성하도록 하는 것은 어떤가요?!

해당 의견에 대해서는 개인적으로 반대하는 의견입니다!
저희가 `HistoryCategory`를 설계했을 때 사전 식물처럼 DB에 기본적으로 제공되고있는 값이라고 생각하고 해당 엔티티를 구성했습니다.
하지만 다음 상황에서 사전 식물과 차별점이 존재합니다.

- HistoryType은 unique값이다.
- HistoryType은 Enum이다.

위 두가지 조건에 의해 테스트 내부에서 해당 도메인 규칙을 계속하여 신경쓰는 부분이 테스트 생산성을 저하할것이라는 생각이 들었습니다.
HistoryCategory에 대해서는 setUp에서 초기화하되 historyCategorySupport에서 조회 로직을 제공하는 방향은 어떻게 생각하시나요?

```java
public class HistoryCategorySupport {

    private final HistoryCategoryRepository historyCategoryRepository;

    public HistoryCategory findByHistoryType(HistoryType type) {
        return historyCategoryRepository.findByHistoryType(type).orElseGet(null);
    }
...
}
```

**Reply by Kim0914** [reply]
오 좋아요 ~
HistoryCategory를 설계했을 때 사전 식물처럼 DB에 기본적으로 제공되고있는 값 이라고 가정했기 때문에 생성하면 부자연스럽네요 ㅎㅎ
historyCategorySupport에서 조회 로직을 제공하는 방향 -> 찬성합니다 😉

**Comment by Choi-JJunho** [comment]
저는 2번 `통합테스트의 DB 초기화 전략을 인수테스트와 동일하게 DatabaseClearExtension을 이용하는 방식을 적용한다.`를 선호합니다.
통합테스트도 `@SpringBootTest`로 모든 컨텍스트를 띄우는 테스트인만큼 DB 초기화를 동일하게 수행해줘도 문제가 없다고 생각했습니다.

인수테스트의 AfterEach에 DB초기화 구문을 작성하면 DB 데이터 초기화를 할 때 불필요한 중복작업이 발생한다는 문제점도 생각해볼 수 있을 것 같아요

```text
BeforeEach - DB 초기화
인수테스트 수행
AfterEach - DB 초기화

[인수테스트 사이에서 불필요한 초기화가 일어남]

BeforeEach - DB 초기화
인수테스트 수행
AfterEach - DB 초기화
```
---

**Reply by Kim0914** [reply]
문득 든 생각인데 `AfterAll`을 사용해볼 수 있지 않을까요?!

**Reply by Choi-JJunho** [reply]
AfterAll이면 제가 우려하고있던 문제를 해결할 수 있어보이네요!

AfterAll을 사용하는 방향이라면 2번 방법보다 괜찮은 방법이라는 생각이 듭니다 👍

**Comment by rawfishthelgh** [comment]
`AfterAll` 어노테이션을 사용하는 방법에는 동의합니다!
주노가 말했던 것 처럼 `AfterEach` 사용 시
`AfterEach`이후  `BeforeEach`가 작동하는 사이에 불필요한 중복 작업의 발생을 줄일 수 있다고 생각이 드네요.

다만 현재 `DatabaseClearExtension`클래스가 테스트에서 격리작업을 담당하고 있는 만큼 
책임분리를 명확하게 하고싶다면 `AfterAllCallback` 인터페이스를 implemens 하여 `afterAll`메소드를 오버라이딩한 격리 작업으로 수행하는 것도 괜찮아 보입니다!
이 경우 1번과 2번을 어느 정도 함께 가져갈 수 있겠네요

**Comment by Kim0914** [comment, marked as answer]
- `AfterAll`을 활용해 인수 테스트가 모두 종료되면 데이터베이스를 클리어한다.
- `historyCategorySupport`에서 null인 경우 생성하지 않고 조회 로직을 제공한다.

위 2가지로 정리할 수 있겠네요 !

**Reply by Choi-JJunho** [reply]
다음 PR에 해당내용 반영했습니다

- https://github.com/woowacourse-teams/2023-pium/pull/246
